website/server/libs/groupTasks.js

```diff
@@ -1,24 +1,68 @@
 import * as Tasks from '../models/task';
+import {model as Groups} from '../models/group';
+import {model as Users} from '../models/user/index';
 
 const SHARED_COMPLETION = {
   default: 'recurringCompletion',
   single: 'singleCompletion',
   every: 'allAssignedCompletion',
 };
 
-async function _completeMasterTask (masterTask) {
-  masterTask.completed = true;
+// @TODO Group task scoring
+async function _completeOrUncompleteMasterTask (masterTask, completed) {
+  masterTask.completed = completed;
   await masterTask.save();
 }
 
-async function _deleteUnfinishedTasks (groupMemberTask) {
-  await Tasks.Task.deleteMany({
-    'group.taskId': groupMemberTask.group.taskId,
-    $and: [
-      {userId: {$exists: true}},
-      {userId: {$ne: groupMemberTask.userId}},
-    ],
-  }).exec();
+async function _updateAssignedUsersTasks (masterTask, groupMemberTask) {
+  if (groupMemberTask.type === 'todo') {
+    if (groupMemberTask.completed) {
+      // The task was done by one person and is removed from others' lists
+      await Tasks.Task.deleteMany({
+        'group.taskId': groupMemberTask.group.taskId,
+        $and: [
+          {userId: {$exists: true}},
+          {userId: {$ne: groupMemberTask.userId}},
+        ],
+      }).exec();
+    } else {
+      // The task was uncompleted by the group member and should be recreated for assignedUsers
+      let group = await Groups.findById(masterTask.group.id).exec();
+      let userList = [];
+      masterTask.group.assignedUsers.forEach(userId => {
+        let query = {_id: userId};
+        userList.push(query);
+      });
+      await Users.find({
+        $or: userList,
+      }).then(async assignedUsers => {
+        for (const assignedUser of assignedUsers) {
+          let promises = [];
+          promises.push(group.syncTask(masterTask, assignedUser));
+          promises.push(group.save());
+          await Promise.all(promises);
+        }
+      });
+    }
+  } else {
+    // Complete or uncomplete the task on other users' lists
+    await Tasks.Task.find({
+      'group.taskId': groupMemberTask.group.taskId,
+      $and: [
+        {userId: {$exists: true}},
+        {userId: {$ne: groupMemberTask.userId}}
+      ]}).then(tasks => {
+        tasks.forEach(task => {
+          // Adjust the task's completion to match the groupMemberTask
+          // Completed or notDue dailies have little effect at cron (MP replenishment for completed dailies)
+          // This maintain's the user's streak without scoring the task if someone else completed the task
+          // If no assignedUser completes the due daily, all users lose their streaks at their cron
+          // @TODO Should we break their streaks or increase their value to encourage competition for the daily?
+          task.completed = groupMemberTask.completed;
+          task.save();
```

Comment:
async ops and array.forEach don't really work well together. You'll want to use a map that return a promise. So for example

```
const tasks = await Tasks.Task.find(...)
await Promise.all(tasks.map(t => {
    ... do stuff with each task...
    return task.save() // return a promise that can be awaited
}))

Reply:
Elegant use of `await Promise.all(tasks.map())`! Thanks!
It also makes a clearer case for not using the `.then()`, though I'm still curious in the thread above.

Reply:
In the end, I used that pattern for re-adding "uncompleted" todos, but optimized this block far more by using mongoose's `updateMany()` instead of `find`ing Users, modifying them, and `save`ing them.